python/meterpreter/meterpreter.py

```diff
@@ -785,12 +786,18 @@ def _read1(self):
             return bytes()
 
     def run(self):
-        byte = self._read1()
-        while len(byte):
-            self.data_lock.acquire()
-            self.data += byte
-            self.data_lock.release()
+        try:
             byte = self._read1()
+            while len(byte) > 0:
+                self.data_lock.acquire()
+                self.data += byte
+                self.data_lock.release()
+                byte = self._read1()
+        finally:
+            self._is_reading = False
```

Comment:
I am not sure of all the possible scenarios for an error that would cause this try block to abort, is it possible and what happens, if an error is raised during `+=`? Would the lock fail to be released?  Would clearing `_is_reading` in the `finally` block end up in a condition where the lock would block some other action or cause a deadlock?

Reply:
I added the `finally` mostly for defensive coding. Worst case scenario if something raises when the thread runs, and `_is_reading` is left as True, the channel would be kept alive indefinitely unless explicitly closed by a user.

There may be existing race conditions around acquiring/releasing this lock under error scenarios, hopefully the future meterpreter test suite will highlight that possibility if so.